File: results/1.0/ELF/elf-gpt-4-turbo.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

// Define sizes and magic numbers
#define EI_NIDENT 16
#define EI_MAG0 0x7F
#define EI_MAG1 'E'
#define EI_MAG2 'L'
#define EI_MAG3 'F'

// ELF identification indexes
enum {
    EI_CLASS = 4,
    EI_DATA = 5,
    EI_VERSION = 6,
    EI_OSABI = 7,
    EI_ABIVERSION = 8
};

// e_type possible values
enum {
    ET_NONE = 0,
    ET_REL = 1,
    ET_EXEC = 2,
    ET_DYN = 3,
    ET_CORE = 4
};

// e_machine possible values
enum {
    EM_NONE = 0,
    EM_X86_64 = 62
};

// Function declarations
static void parse_elf(const char *filename);

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ELF file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_elf(argv[1]);
    return 0;
}

static void parse_elf(const char *filename) {
    // Open the ELF file
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    // Move file pointer to the end and get size
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    // Read the entire file into memory
    unsigned char *buffer = malloc(file_size);
    if (!buffer) {
        fclose(file);
        fprintf(stderr, "Failed to allocate memory\n");
        exit(EXIT_FAILURE);
    }
    fread(buffer, 1, file_size, file);
    fclose(file);

    // Define ELF header parsers
    HParser *magic = h_sequence(h_int8(), h_ch(EI_MAG1), h_ch(EI_MAG2), h_ch(EI_MAG3), NULL);
    HParser *ident = h_sequence(magic,
                                h_int8(),  // EI_CLASS
                                h_int8(),  // EI_DATA
                                h_int8(),  // EI_VERSION
                                h_int8(),  // EI_OSABI
                                h_int8(),  // EI_ABIVERSION
                                h_ignore(h_repeat_n(h_int8(), 7)), // EI_PAD
                                NULL);
    HParser *e_type = h_int16();
    HParser *e_machine = h_int16();
    HParser *e_version = h_int32();
    HParser *e_entry = h_int64();
    HParser *e_phoff = h_int64();
    HParser *e_shoff = h_int64();
    HParser *e_flags = h_int32();
    HParser *e_ehsize = h_int16();
    HParser *e_phentsize = h_int16();
    HParser *e_phnum = h_int16();
    HParser *e_shentsize = h_int16();
    HParser *e_shnum = h_int16();
    HParser *e_shstrndx = h_int16();

    HParser *elf_header = h_sequence(ident, e_type, e_machine, e_version,
                                     e_entry, e_phoff, e_shoff, e_flags,
                                     e_ehsize, e_phentsize, e_phnum, 
                                     e_shentsize, e_shnum, e_shstrndx, NULL);

    // Parse the ELF header
    HParseResult *result = h_parse(elf_header, buffer, file_size);
    if (result) {
        printf("ELF parsed successfully.\n");
        h_pprint(stdout, result->ast, 0, 3);
        h_parse_result_free(result);
    } else {
        printf("Failed to parse ELF.\n");
    }

    // Cleanup
    free(buffer);
}

File: results/1.0/ELF/elf-gpt-4-turbo.rs
use nom::{
    number::complete::{le_u32, le_u16, le_u64, le_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

const EI_NIDENT: usize = 16;

#[derive(Debug)]
struct ElfHeader {
    e_ident: [u8; EI_NIDENT],
    e_type: u16,
    e_machine: u16,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, e_ident) = nom::bytes::complete::take(EI_NIDENT)(input)?;
    let (input, e_type) = le_u16(input)?;
    let (input, e_machine) = le_u16(input)?;
    let (input, e_version) = le_u32(input)?;
    let (input, e_entry) = le_u64(input)?;
    let (input, e_phoff) = le_u64(input)?;
    let (input, e_shoff) = le_u64(input)?;
    let (input, e_flags) = le_u32(input)?;
    let (input, e_ehsize) = le_u16(input)?;
    let (input, e_phentsize) = le_u16(input)?;
    let (input, e_phnum) = le_u16(input)?;
    let (input, e_shentsize) = le_u16(input)?;
    let (input, e_shnum) = le_u16(input)?;
    let (input, e_shstrndx) = le_u16(input)?;

    let header = ElfHeader {
        e_ident: {
            let mut ident = [0u8; EI_NIDENT];
            ident.copy_from_slice(e_ident);
            ident
        },
        e_type,
        e_machine,
        e_version,
        e_entry,
        e_phoff,
        e_shoff,
        e_flags,
        e_ehsize,
        e_phentsize,
        e_phnum,
        e_shentsize,
        e_shnum,
        e_shstrndx,
    };
    Ok((input, header))
}

fn read_elf_file(path: &str) -> io::Result<Vec<u8>> {
    let mut file = File::open(path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;
    Ok(buffer)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <ELF file path>", args[0]);
        std::process::exit(1);
    }

    let path = &args[1];
    match read_elf_file(path) {
        Ok(bytes) => match parse_elf_header(&bytes) {
            Ok((_, header)) => {
                println!("{:#?}", header);
            }
            Err(e) => {
                eprintln!("Failed to parse ELF header: {:?}", e);
                std::process::exit(1);
            }
        },
        Err(e) => {
            eprintln!("Failed to read file: {:?}", e);
            std::process::exit(1);
        }
    }
}